TVHeadEnd/AccessTicketHandler.cs

```diff
@@ -0,0 +1,141 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using TVHeadEnd.HTSP;
+using TVHeadEnd.HTSP_Responses;
+using TVHeadEnd.TimeoutHelper;
+
+namespace TVHeadEnd;
+
+public class AccessTicketHandler
+{
+    private readonly ILogger<AccessTicketHandler> _logger;
+
+    private readonly HTSConnectionHandler _htsConnectionHandler;
+    private readonly string _ticketItemType;
+    private readonly TimeSpan _requestTimeout;
+    private readonly int _requestRetries;
+    private readonly TimeSpan _ticketLifeSpan;
+
+    private volatile int _ticketIdSequence;
+
+    public enum TicketType : byte { Channel, Recording };
+
+    public record Ticket
+    {
+        public string Id { get; init; }
+        public string Path { get; init; }
+        public string TicketParam { get; init; }
+        public string Url => $"{Path}?ticket={TicketParam}";
+        public DateTime Expires { get; init; }
+    }
+
+    private readonly ConcurrentDictionary<string, Task<Ticket>> _ticketCache = new();
+
+    internal AccessTicketHandler(
+        ILoggerFactory loggerFactory, HTSConnectionHandler htsConnectionHandler,
+        TimeSpan requestTimeout, int requestRetries, TimeSpan ticketLifeSpan, TicketType ticketType)
+    {
+        _logger = loggerFactory.CreateLogger<AccessTicketHandler>();
+        _htsConnectionHandler = htsConnectionHandler;
+        _requestTimeout = requestTimeout;
+        _requestRetries = requestRetries;
+        _ticketLifeSpan = ticketLifeSpan;
+
+        _ticketItemType = ticketType switch
+        {
+            TicketType.Channel => "channelId",
+            TicketType.Recording => "dvrId",
+            _ => throw new ArgumentException("undefined ticketType")
+        };
+    }
+
+    public async Task<Ticket> GetTicket(string itemId, CancellationToken cancellationToken)
+    {
+        var now = DateTime.UtcNow;
+        Ticket ticket = null;
+
+        while (_ticketCache.TryGetValue(itemId, out var ticketTask))
+        {
+            ticket = await ticketTask;
+            if (ticket.Expires > now)
+            {
+                return ticket; // non-expired ticket from cache
+            }
+
+            _logger.LogDebug("[TVHclient] AccessTicketHandler.GetAccessTicket: Cache expired for {ItemType}={ItemId}. Revalidating ticket (#{TicketId})", _ticketItemType, itemId, ticket.Id);
+            _ticketCache.TryRemove(new KeyValuePair<string, Task<Ticket>>(itemId, ticketTask));
+        }
+
+        return await _ticketCache.GetOrAdd(itemId, _ => GetTicketRecord(itemId, cancellationToken, ticket, now));
+    }
+
+    private Task<Ticket> GetTicketRecord(string itemId, CancellationToken cancellation, Ticket currentRecord, DateTime now)
+    {
+        return RequestTicket(itemId, cancellation).ContinueWith(ticketTask =>
```

Comment:
I think I would prefer removing the `ContinueWith` and just have these calls be awaited

Reply:
Might be wrong but can’t use await here since the method can’t be marked async (it is used in a closure)